### DISCUSSION TITLE
PyInstaller packages 340 shared objects in Linux. (Total binary size 854Mb)

### DISCUSSION THREAD

**Discussion by olumide-x** [pyinstaller/pyinstaller#8265] [Help]
I am aware that the question of Linux binary sizes is a recurring one but I have tried the recommendation to [run PyInstaller in a virtual env](https://stackoverflow.com/a/47810133/181783) and its made no difference.

The application itself is a modest 15Mb unfortunately it's mapped a staggering number of shared objects into its address space, as shown when I run the command

`ll /proc/<PID>/map_files/ | grep "\.so" | cut -d " " -f 11 | sort | uniq`

or

`lsof -p <PID> | grep "\.so" | wc -l`

I see something like [this](https://pastebin.com/hGN8NWfZ).

For example 87 scipy shared objects were gathered from scipy even though the application imports just one scipy module

`import scipy.stats.distributions`

Is there a way to cull the excessive mapping of shared objects?

**Comment by rokm** [comment]
And how many shared objects are mapped when you run your application unfrozen?

**Comment by olumide-x** [comment]
> And how many shared objects are mapped when you run your application unfrozen?

282, so it looks like its a Python on Linux issue. Please close if you agree.

**Reply by olumide-x** [reply]
I had a look at the memory map of the process on Windows on MYSYS. It links just 62 non-Windows dlls.

`grep "\.dll" /proc/<PID>/maps | grep -ve "/c/Windows/" | cut -d " " -f 8 | sort | uniq | wc -l`

**Reply by rokm** [reply]
> I had a look at the memory map of the process on Windows on MYSYS. It links just 62 non-Windows dlls.
> 
> `grep "\.dll" /proc/<PID>/maps | grep -ve "/c/Windows/" | cut -d " " -f 8 | sort | uniq | wc -l`

I fail to see how this is relevant to behavior of unfrozen and frozen script under linux?

**Reply by olumide-x** [reply]
I was speculating that this might be the reason why PyInstalller-built Windows binaries are not as bloated as PyInstaller-built Linux binaries.

**Reply by bwoodsend** [reply]
That's more because Linux is decentralised so that it has less universally available libraries than the other OSs. On Linux, we can bank on libc being installed and that's about it. Pretty much everything else has to be bundled in.

PyInstaller doesn't actually make much sense on Linux. I don't understand why it's our most commonly used platform.

**Reply by olumide-x** [reply]
What I have since learned is that all those (needless) shared objects are linked during while the runtime is parsing the __init__.py files. Sadly they are not unmapped after initializing the module.

I wonder what your other users are doing to keep their binary sizes small.

**Reply by bwoodsend** [reply]
If they're being loaded, then they're not needless. Your program will crash without them.

> I wonder what your other users are doing to keep their binary sizes small.

Some of them will be using the relatively hidden [exclude_system_libraries](https://pyinstaller.org/en/v6.0.0/spec-files.html?highlight=exclude_system_libraries#posix-specific-options) option at the expense of portability.
